src/ops/leaf.rs

```diff
@@ -49,6 +52,59 @@ pub enum LeafOp {
         /// The extensions which we're adding to the inputs
         new_extension: ExtensionId,
     },
+    /// Fixes some [TypeParam]s of a polymorphic type by providing [TypeArg]s
+    ///
+    /// [TypeParam]: crate::types::type_param::TypeParam
+    TypeApply {
+        /// The type and args, plus a cache of the resulting type
+        ta: TypeApplication,
+    },
+}
+
+/// Records details of an application of a [PolyFuncType] to some [TypeArg]s
+/// and the result (a less-, but still potentially-, polymorphic type).
+#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct TypeApplication {
+    input: PolyFuncType,
+    args: Vec<TypeArg>,
+    output: PolyFuncType, // cached
+}
+
+impl TypeApplication {
+    /// Checks that the specified args are correct for the [TypeParam]s of the polymorphic input.
+    /// Note the extension registry is required here to recompute [Type::least_upper_bound]s.
+    ///
+    /// [TypeParam]: crate::types::type_param::TypeParam
+    pub fn try_new(
+        input: PolyFuncType,
+        args: impl Into<Vec<TypeArg>>,
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<Self, SignatureError> {
+        let args = args.into();
+        // Should we require >=1 `arg`s here? Or that input declares >=1 params?
+        // At the moment we allow an identity TypeApply on a monomorphic function type.
```

Comment:
Mhm I don't have a strong preference here. Allowing identity TypeApplies wouldn't hurt but I also don't see a reason why you would want to use them

Reply:
I suspect allowing identity type applies will make certain rewrites and simplifications easier to express, can always be removed in cleanup phases

Reply:
Makes sense yes